File: colorLib/src/main/java/org/treebolic/colors/view/ColorPanelView.java
/*
 * Copyright (c) 2019. Bernard Bou <1313ou@gmail.com>
 */

package org.treebolic.colors.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import org.treebolic.colors.drawable.AlphaPatternDrawable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * This class draws a panel which which will be filled with a value which can be set. It can be used to show the currently selected value which you will get
 * from the {@link ColorPickerView}.
 *
 * @author Daniel Nilsson
 */
public class ColorPanelView extends View
{
	/**
	 * The width in pixels of the border surrounding the value panel.
	 */
	private final static float BORDER_WIDTH_PX = 1;

	/**
	 * Density
	 */
	private static float density = 1f;

	/**
	 * Border value
	 */
	private int borderColor = 0xff6E6E6E;

	/**
	 * Color
	 */
	private int color = 0xff000000;

	/**
	 * Back paint
	 */
	static private final Paint BACK_PAINT = new Paint();

	static
	{
		BACK_PAINT.setColor(Color.WHITE);
	}

	/**
	 * Draw paint
	 */
	static private final Paint DRAW_PAINT = new Paint();

	static
	{
		DRAW_PAINT.setColor(Color.GRAY);
	}

	/**
	 * Border paint
	 */
	private Paint borderPaint;

	/**
	 * Color paint
	 */
	private Paint colorPaint;

	/**
	 * Drawing rect
	 */
	private RectF drawingRect;

	/**
	 * Color rect
	 */
	private RectF colorRect;

	/**
	 * Alpha pattern
	 */
	private AlphaPatternDrawable alphaPattern;

	/**
	 * IsNull
	 */
	private boolean isNull;

	/**
	 * IsIllegal
	 */
	private boolean isCrossed;

	/**
	 * Constructor
	 *
	 * @param context context
	 */
	public ColorPanelView(@NonNull final Context context)
	{
		this(context, null);
	}

	/**
	 * Constructor
	 *
	 * @param context context
	 * @param attrs   attributes
	 */
	public ColorPanelView(@NonNull final Context context, final AttributeSet attrs)
	{
		this(context, attrs, 0);
	}

	/**
	 * Constructor
	 *
	 * @param context  context
	 * @param attrs    attributes
	 * @param defStyle style
	 */
	public ColorPanelView(@NonNull final Context context, final AttributeSet attrs, final int defStyle)
	{
		super(context, attrs, defStyle);
		init(context);
	}

	/**
	 * Common init
	 *
	 * @param context context
	 */
	private void init(@NonNull final Context context)
	{
		this.colorPaint = new Paint();
		this.borderPaint = new Paint();
		ColorPanelView.density = context.getResources().getDisplayMetrics().density;
	}

	@Override
	protected void onDraw(@NonNull final Canvas canvas)
	{
		// border
		//noinspection ConstantConditions
		if (ColorPanelView.BORDER_WIDTH_PX > 0)
		{
			this.borderPaint.setColor(this.borderColor);
			canvas.drawRect(this.drawingRect, this.borderPaint);
		}

		// crossed
		if (this.isCrossed)
		{
			canvas.drawRect(this.colorRect, BACK_PAINT);

			canvas.drawLine(this.drawingRect.left, this.drawingRect.top, this.drawingRect.right, this.drawingRect.bottom, DRAW_PAINT);
			canvas.drawLine(this.drawingRect.right, this.drawingRect.top, this.drawingRect.left, this.drawingRect.bottom, DRAW_PAINT);
			return;
		}

		// pattern
		if (this.alphaPattern != null)
		{
			this.alphaPattern.draw(canvas);
		}

		// value
		if (!this.isNull)
		{
			this.colorPaint.setColor(this.color);
			final RectF rect = this.colorRect;
			canvas.drawRect(rect, this.colorPaint);
		}
	}

	@Override
	protected void onMeasure(final int widthMeasureSpec, final int heightMeasureSpec)
	{
		final int width = MeasureSpec.getSize(widthMeasureSpec);
		final int height = MeasureSpec.getSize(heightMeasureSpec);
		setMeasuredDimension(width, height);
	}

	@Override
	protected void onSizeChanged(final int w, final int h, final int oldW, final int oldH)
	{
		super.onSizeChanged(w, h, oldW, oldH);

		this.drawingRect = new RectF();
		this.drawingRect.left = getPaddingLeft();
		this.drawingRect.right = w - getPaddingRight();
		this.drawingRect.top = getPaddingTop();
		this.drawingRect.bottom = h - getPaddingBottom();
		setUpColorRect();
	}

	/**
	 * Set up value rectangle
	 */
	private void setUpColorRect()
	{
		final RectF dRect = this.drawingRect;
		final float left = dRect.left + ColorPanelView.BORDER_WIDTH_PX;
		final float top = dRect.top + ColorPanelView.BORDER_WIDTH_PX;
		final float bottom = dRect.bottom - ColorPanelView.BORDER_WIDTH_PX;
		final float right = dRect.right - ColorPanelView.BORDER_WIDTH_PX;

		this.colorRect = new RectF(left, top, right, bottom);
		this.alphaPattern = new AlphaPatternDrawable((int) (5 * ColorPanelView.density));
		this.alphaPattern.setBounds(Math.round(this.colorRect.left), Math.round(this.colorRect.top), Math.round(this.colorRect.right), Math.round(this.colorRect.bottom));
		this.isCrossed = false;
	}

	/**
	 * Set value
	 *
	 * @param color may be null
	 */
	public void setValue(@Nullable final Integer color)
	{
		if (color == null)
		{
			this.isNull = true;
			this.color = 0x00ffffff;
			invalidate();
			return;
		}
		setColor(color);
	}

	/**
	 * Set the value that should be shown by this view.
	 *
	 * @param color value
	 */
	public void setColor(final int color)
	{
		this.isNull = false;
		this.color = color;
		invalidate();
	}

	/**
	 * Get the value currently shown by this view.
	 *
	 * @return value
	 */
	public int getColor()
	{
		return this.color;
	}

	/**
	 * Set the value of the border surrounding the panel.
	 *
	 * @param color border value
	 */
	public void setBorderColor(@SuppressWarnings("SameParameterValue") final int color)
	{
		this.borderColor = color;
		invalidate();
	}

	/**
	 * Get the value of the border surrounding the panel.
	 *
	 * @return border value
	 */
	public int getBorderColor()
	{
		return this.borderColor;
	}

	/**
	 * Set crossed flag
	 *
	 * @param isCrossed whether the display is crossed
	 */
	public void setCrossed(final boolean isCrossed)
	{
		this.isCrossed = isCrossed;
	}
}

File: colorLib/src/main/java/org/treebolic/colors/view/ColorPanelView.kt
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */
package org.treebolic.colors.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import org.treebolic.colors.drawable.AlphaPatternDrawable
import kotlin.math.roundToInt

/**
 * This class draws a panel which which will be filled with a value which can be set. It can be used to show the currently selected value which you will get
 * from the [ColorPickerView].
 *
 * @param context  context
 * @param attrs    attributes
 * @param defStyle style
 *
 * @author Daniel Nilsson
 * @author Bernard
 */
class ColorPanelView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : View(context, attrs, defStyle) {

    /**
     * Border value
     */
    private var borderColor = -0x919192

    /**
     * Color, the value that should be shown by this view.
     */
    var color: Int = -0x1000000
        set(newColor) {
            isNull = false
            field = newColor
            invalidate()
        }

    /**
     * Border paint
     */
    private lateinit var borderPaint: Paint

    /**
     * Color paint
     */
    private lateinit var colorPaint: Paint

    /**
     * Drawing rect
     */
    private var drawingRect: RectF? = null

    /**
     * Color rect
     */
    private var colorRect: RectF? = null

    /**
     * Alpha pattern
     */
    private var alphaPattern: AlphaPatternDrawable? = null

    /**
     * IsNull
     */
    private var isNull = false

    /**
     * IsIllegal
     */
    var isCrossed: Boolean = false

    init {
        init(context)
    }

    /**
     * Common init
     *
     * @param context context
     */
    private fun init(context: Context) {
        colorPaint = Paint()
        borderPaint = Paint()
        density = context.resources.displayMetrics.density
    }

    override fun onDraw(canvas: Canvas) {

        // border
        if (BORDER_WIDTH_PX > 0) {
            borderPaint.color = borderColor
            canvas.drawRect(drawingRect!!, borderPaint)
        }

        // crossed
        if (isCrossed) {
            canvas.drawRect(colorRect!!, BACK_PAINT)

            canvas.drawLine(drawingRect!!.left, drawingRect!!.top, drawingRect!!.right, drawingRect!!.bottom, DRAW_PAINT)
            canvas.drawLine(drawingRect!!.right, drawingRect!!.top, drawingRect!!.left, drawingRect!!.bottom, DRAW_PAINT)
            return
        }

        // pattern
        if (alphaPattern != null) {
            alphaPattern!!.draw(canvas)
        }

        // value
        if (!isNull) {
            colorPaint.color = color
            val rect = colorRect
            canvas.drawRect(rect!!, colorPaint)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = MeasureSpec.getSize(widthMeasureSpec)
        val height = MeasureSpec.getSize(heightMeasureSpec)
        setMeasuredDimension(width, height)
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)

        drawingRect = RectF()
        drawingRect!!.left = paddingLeft.toFloat()
        drawingRect!!.right = (w - paddingRight).toFloat()
        drawingRect!!.top = paddingTop.toFloat()
        drawingRect!!.bottom = (h - paddingBottom).toFloat()
        setUpColorRect()
    }

    /**
     * Set up value rectangle
     */
    private fun setUpColorRect() {
        val dRect = drawingRect
        val left = dRect!!.left + BORDER_WIDTH_PX
        val top = dRect.top + BORDER_WIDTH_PX
        val bottom = dRect.bottom - BORDER_WIDTH_PX
        val right = dRect.right - BORDER_WIDTH_PX

        colorRect = RectF(left, top, right, bottom)
        alphaPattern = AlphaPatternDrawable((5 * density).toInt())
        alphaPattern!!.setBounds(colorRect!!.left.roundToInt(), colorRect!!.top.roundToInt(), colorRect!!.right.roundToInt(), colorRect!!.bottom.roundToInt())
        isCrossed = false
    }

    /**
     * Set value
     *
     * @param newValue may be null
     */
    fun setValue(newValue: Int?) {
        if (newValue == null) {
            isNull = true
            color = 0x00ffffff
            invalidate()
            return
        }
        color = newValue
    }

    /**
     * Set the value of the border surrounding the panel.
     *
     * @param color border value
     */
    fun setBorderColor(color: Int) {
        borderColor = color
        invalidate()
    }

    companion object {

        /**
         * The width in pixels of the border surrounding the value panel.
         */
        private const val BORDER_WIDTH_PX = 1f

        /**
         * Density
         */
        private var density = 1f

        /**
         * Back paint
         */
        private val BACK_PAINT = Paint().apply {
            color = Color.WHITE
        }

        /**
         * Draw paint
         */
        private val DRAW_PAINT = Paint().apply {
            color = Color.GRAY
        }
    }
}
